### ISSUE TITLE
Unable to hold dimensionless values as Quantity'es

### ISSUE THREAD

**Issue by cnaak** [PainterQubits/Unitful.jl#283]
`Unitful.jl` has definitions of both Units and Dimensions for dimensionless quantities, namely:

- `Unitful.FreeUnits{(),NoDims,nothing}`, for Units, and
- `NoDims`,for Dimensions.

One expects that dimensionless quantities can thus be held in `Quantity` types, such as `Quantity{Float64,NoDims,Unitful.FreeUnits{(),NoDims,nothing}}`; however, this is *not* the case **as dimensionless amounts automatically degenerate to POD's (plain old data)**:

```julia-repl
julia> x = 1.0u"kg"; typeof(x)
Quantity{Float64,𝐌,Unitful.FreeUnits{(kg,),𝐌,nothing}}

julia> typeof(x / u"kg")
Float64
```

I have a parametric type with a single `Quantity` field. One of the parameters is the physical base for the quantity—whether mass-base or molar-base, as commonly found in Thermodynamics—that fails to instantiate when units cancel, because a `Float64` cannot be "converted" to a `Quantity{Float64,NoDims,Unitful.FreeUnits{(),NoDims,nothing}}`.

As a **MWE**:

```julia
abstract type BASE end
abstract type SY <: BASE end # The system base
abstract type MA <: BASE end # The mass base
abstract type MO <: BASE end # The molar base

uSY = u"kg"          # System mass in kg
uMA = uSY / u"kg"    # Mass fraction in kg/kg
uMO = uSY / u"kmol"  # Molecular mass in kg/kmol
dSY = dimension(uSY)
dMA = dimension(uMA)
dMO = dimension(uMO)
struct mAmt{B<:Union{SY,MA,MO}}
    amt::Union{Quantity{Float64,dSY},Quantity{Float64,dMA},Quantity{Float64,dMO}}
    # amt can be meaningfully dimensionless
    mAmt(x::Float64, ::Type{SY}) = new{SY}(x * uSY)
    mAmt(x::Float64, ::Type{MA}) = new{MA}(x * uMA)
    mAmt(x::Float64, ::Type{MO}) = new{MO}(x * uMO)
end
```

Now lets try it:

```julia-repl
julia> mAmt(1.0, SY)
mAmt{SY}(1.0 kg)

julia> mAmt(1.0, MO)
mAmt{MO}(1.0 kg kmol^-1)

julia> mAmt(1.0, MA)
ERROR: MethodError: no method matching Union{Quantity{Float64,𝐌,U} where U, Quantity{Float64,NoDims,U} whe
re U, Quantity{Float64,𝐌*𝐍^-1,U} where U}(::Float64)
[...]
```

In the real application, the MWE above is created by metaprogramming, not only for mass, but for many different quantities, and the issue being reported is what's preventing the design to be uniform throughout.

**Comment by ajkeller34** [comment]
`Quantity`s work pretty well most of the time, but people don't always write generic code, and so it happens that often it is very convenient to have `Float64`s instead of `Quantity`s where possible. This was an intentional design choice. 

There's nothing preventing you from constructing dimensionless, unitless types explicitly, so you could for example do:

```
qty(x) = Quantity{Float64, dimension(x), typeof(unit(x))}(x)
struct mAmt{B<:Union{SY,MA,MO}}
    amt::Union{Quantity{Float64,dSY},Quantity{Float64,dMA},Quantity{Float64,dMO}}
    # amt can be meaningfully dimensionless
    mAmt(x::Float64, ::Type{SY}) = new{SY}(qty(x * uSY))
    mAmt(x::Float64, ::Type{MA}) = new{MA}(qty(x * uMA))
    mAmt(x::Float64, ::Type{MO}) = new{MO}(qty(x * uMO))
end
```

That being said, I don't think that's great design, in part because `amt` is not concretely typed. You might instead try something like:

```
struct mAmt2{B<:Union{SY,MA,MO}, T<:Number}
    amt::T
end
mAmt2{B}(x::T) where {B, T} = mAmt2{B,T}(x)
mAmt2(x, ::Type{SY}) = mAmt2{SY}(Float64(x) * uSY)
mAmt2(x, ::Type{MA}) = mAmt2{MA}(Float64(x) * uMA)
mAmt2(x, ::Type{MO}) = mAmt2{MO}(Float64(x) * uMO)
```

By writing the struct in this way, you don't care if `T<:Quantity` or `T<:Float64`.

**Comment by cnaak** [comment]
> `qty(x) = Quantity{Float64, dimension(x), typeof(unit(x))}(x)`

Thank you!! This beautifully solves my issue!

**Comment by cnaak** [comment]
Thank you for commenting on the MWE design. _It was very welcomed_; however, I do not follow this part:

> [...] in part because `amt` is not concretely typed [...]

Is that because in the concrete type declaration I used `amt::Union{...}`, i.e., a type `Union` instead of an Abstract supertype?

The type annotation for the `amt` member always makes `amt<:Number`, and the constructors make sure all _instances_ are concretely typed, for instance:

```julia-repl
julia> typeof(mAmt(1.0, SY).amt)
Quantity{Float64,𝐌,Unitful.FreeUnits{(kg,),𝐌,nothing}}
```

**Comment by ajkeller34** [comment]
> Is that because in the concrete mAmt type declaration I used amt::Union{...}, i.e., a type Union instead of an Abstract supertype?

If you declare the type of a struct's field to be an abstract type, or a union, the compiler can only infer so much at compile-time. There's usually, but not always, a performance hit. Small unions may actually be fine now, but it wasn't always this way. See https://julialang.org/blog/2018/08/union-splitting. 

If you consider what I was doing above with `mAmt2`, `T` *could* be an abstract type, but a constructor will usually try to assign a concrete type to that parameter based on the input arguments to the constructor. In this way the compiler can figure out at compile time what the field types are. This is in large part how Unitful is able to remain pretty fast in spite of having units. I'd suggest reading more about parametric types in the Julia docs to get a feel for how this works.

For example, consider the output of this function:

```
julia> g(x) = x.amt
g (generic function with 1 method)
```

If we inspect the function for type stability, we see:
```
julia> @code_warntype g(mAmt(1.0,SY))
Variables
  #self#::Core.Compiler.Const(g, false)
  x::mAmt{SY}

Body::Union{Quantity{Float64,𝐌,U} where U, Quantity{Float64,NoDims,U} where U, Quantity{Float64,𝐌*𝐍^-1,U} where U}
1 ─ %1 = Base.getproperty(x, :amt)::Union{Quantity{Float64,𝐌,U} where U, Quantity{Float64,NoDims,U} where U, Quantity{Float64,𝐌*𝐍^-1,U} where U}
└──      return %1
```
This is a bit hard to follow if you're not used to reading it, but where it says `Body::Union`, that indicates the return type of `g` can only be inferred by the compiler to be a union of different `Quantity` types. If you were to instead try with `mAmt2` as defined above, you'd see:

```
julia> @code_warntype g(mAmt2(1.0,SY))
Variables
  #self#::Core.Compiler.Const(g, false)
  x::mAmt2{SY,Quantity{Float64,𝐌,Unitful.FreeUnits{(kg,),𝐌,nothing}}}

Body::Quantity{Float64,𝐌,Unitful.FreeUnits{(kg,),𝐌,nothing}}
1 ─ %1 = Base.getproperty(x, :amt)::Quantity{Float64,𝐌,Unitful.FreeUnits{(kg,),𝐌,nothing}}
└──      return %1
```
Here, the compiler knows which specific concrete type is returned. 

I think you may have a misconception that the compiler is able to tell which kind of number is stored in the struct based on which constructor is used. There's no link exposed to the type system between the number type and the `B` type parameter. When you're using `typeof` to check that the type is concrete, that's not the same as the compiler being able to infer the type—you're using run-time information. 

Another way to see this is to use the `Test.@inferred` macro:

```
julia> using Test

julia> h(x) = typeof(x.amt)
h (generic function with 1 method)

julia> @inferred h(mAmt(1.0, SY))
ERROR: return type Type{Quantity{Float64,𝐌,Unitful.FreeUnits{(kg,),𝐌,nothing}}} does not match inferred return type Union{Type{Quantity{Float64,𝐌,U}} where U, Type{Quantity{Float64,NoDims,U}} where U, Type{Quantity{Float64,𝐌*𝐍^-1,U}} where U}
Stacktrace:
 [1] top-level scope at none:0

julia> @inferred h(mAmt2(1.0, SY))
Quantity{Float64,𝐌,Unitful.FreeUnits{(kg,),𝐌,nothing}}
```

Finally, let me just say that these are really just examples for you to think about. I wouldn't necessarily suggest doing exactly what I did with `mAmt2`; it can be more work to make parametric types work well and if your code works fine now for your purposes, that's great. I just wanted to make sure you were aware of some of these considerations.

**Comment by cnaak** [comment]
Thank you so much for taking the time to explain all of this with examples and references!

I'm anticipating a refactoring of the code that motivated this thread, as to avoid type stability issues.

**Comment by ajkeller34** [comment]
Sure, you're welcome! Will close the issue for now.
